src/ClientTransactionTask.cc

```diff
@@ -0,0 +1,562 @@
+                                                            // method call.
+        participantList.emplaceAppend<WireFormat::TxParticipant>(
+                key->tableId,
+                static_cast<uint64_t>(key->keyHash),
+                entry->rpcId);
+        participantCount++;
+        nextCacheEntry++;
+    }
+}
+
+/**
+ * Process any decision rpcs that have completed.  Used in the commit method.
+ * Factored out mostly for ease of testing.
+ */
+void
+ClientTransactionTask::processDecisionRpcs()
+{
+    // Process outstanding RPCs.
+    std::list<DecisionRpc>::iterator it = decisionRpcs.begin();
+    for (; it != decisionRpcs.end(); it++) {
+        DecisionRpc* rpc = &(*it);
+
+        if (!rpc->isReady()) {
+            continue;
+        }
+
+        if (rpc->getState() == rpc->FAILED) {
+            // Nothing to do.  Will be retried.
+            TEST_LOG("FAILED");
+        } else if (rpc->responseHeader->status == STATUS_OK) {
+            TEST_LOG("STATUS_OK");
+            for (uint32_t i = 0; i < rpc->reqHdr->participantCount; i++)
+                ramcloud->rpcTracker.rpcFinished(rpc->ops[i]->second.rpcId);
+        } else if (rpc->responseHeader->status == STATUS_UNKNOWN_TABLET) {
+            // Nothing to do.  Will be retried.
+            TEST_LOG("STATUS_UNKNOWN_TABLET");
+        } else {
+            ClientException::throwException(HERE, rpc->responseHeader->status);
+        }
+
+        // Destroy object.
+        it = decisionRpcs.erase(it);
+    }
+}
+
+/**
+ * Process any prepare rpcs that have completed.  Used in the commit method.
+ * Factored out mostly for ease of testing.
+ */
+void
+ClientTransactionTask::processPrepareRpcs()
+{
+    // Process outstanding RPCs.
+    std::list<PrepareRpc>::iterator it = prepareRpcs.begin();
+    for (; it != prepareRpcs.end(); it++) {
+        PrepareRpc* rpc = &(*it);
+
+        if (!rpc->isReady()) {
+            continue;
+        }
+
+        if (rpc->getState() == rpc->FAILED) {
+            // Nothing to do.  Will be retried.
+            TEST_LOG("FAILED");
+        } else if (rpc->responseHeader->status == STATUS_OK) {
+            WireFormat::TxPrepare::Response* respHdr =
+                    rpc->response->getStart<WireFormat::TxPrepare::Response>();
+            if (respHdr->vote != WireFormat::TxPrepare::COMMIT) {
+                decision = WireFormat::TxDecision::ABORT;
+            }
+        } else if (rpc->responseHeader->status == STATUS_UNKNOWN_TABLET) {
+            // Nothing to do.  Will be retried.
+            TEST_LOG("STATUS_UNKNOWN_TABLET");
+        } else {
+            ClientException::throwException(HERE, rpc->responseHeader->status);
+        }
+
+        // Destroy object.
+        it = prepareRpcs.erase(it);
+    }
+}
+
+/**
+ * Send out a decision rpc if not all master have been notified.  Used in the
+ * commit method.  Factored out mostly for ease of testing.
+ */
+void
+ClientTransactionTask::sendDecisionRpc()
```

Comment:
Could you combine this method with sendPrepareRpc by creating a version that is templated on the RPC type?